Sources_v3/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware.swift

```diff
@@ -0,0 +1,56 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// A middleware which updates a channel's read events as websocket events arrive.
+struct ChannelReadUpdaterMiddleware<ExtraData: ExtraDataTypes>: EventMiddleware {
+    let database: DatabaseContainer
+    
+    func handle(event: Event, completion: @escaping (Event?) -> Void) {
+        defer { completion(event) }
+        
+        if let event = event as? MessageReadEvent<ExtraData> {
+            updateReadEvent(for: event.cid, userId: event.userId, lastReadAt: event.readAt)
+            return
+        }
+        
+        if let event = event as? NotificationMarkReadEvent<ExtraData> {
+            updateReadEvent(for: event.cid, userId: event.userId, lastReadAt: event.readAt)
+            return
+        }
+        
+        if let event = event as? NotificationMarkAllReadEvent<ExtraData> {
+            let readsForUser = ChannelReadDTO.load(userId: event.userId, context: database.viewContext)
+            database.write({ _ in
+                readsForUser.forEach { read in
+                    read.lastReadAt = event.readAt
+                    read.unreadMessageCount = 0
+                }
+            }, completion: { error in
+                if let error = error {
+                    log.error("Failed to update channel reads for userId \(event.userId), error: \(error)")
+                }
```

Comment:
... this is the point where you should call `completion(event)`

Reply:
Added in b65ae9e402eac93dfee89c9c09afa10935dd4c21